### PR TITLE
Docker improvements

### PR DESCRIPTION
**Pull request by andyneff**
Vastly cleaned up docker directory and files

Removed need to rebuild entire environment just to get dependencies
Normalized all rpm names so they follow the same pattern between OSes
Added 32 bit cross compile support
Move GPG agent from host to docker, to support any boot2docker host now. 
Fixed file ownership issue
Enabled integration test on rpm build
Cleaned up README

### CODE REVIEW

File: docker/docker+.bsh [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+set -eu
+
+if [[ $# == 0 ]]; then
+  FILES=(-)
+else
+  FILES=("${@}")
+fi
+
+CUR_DIR=$(basename ${BASH_SOURCE[0]})
```
Comment (technoweenie) [inline docker/docker+.bsh]: Should this be `dirname` instead? Otherwise this is just `docker+.bsh` when run on my mac.

File: docker/docker+.bsh [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  while IFS='' read -r LINE || [ -n "${LINE}" ]; do
+
+    #My template substitution
+    while [[ "$LINE" =~ (\[\{)([A-Za-z0-9_]*)(\}\]) ]]; do
+      LINE=${LINE/"${BASH_REMATCH[0]}"/${!BASH_REMATCH[2]-}}
+    done
+    
+    #My SOURCE substitution
+    if echo $LINE | grep -q ^SOURCE; then
+      SOURCE=$(echo "${LINE}" | sed -r 's|^SOURCE\s+(\S*)\s*|\1|')
```
Comment (technoweenie) [inline docker/docker+.bsh]: This line blows up on the mac, which apparently has a really old version of sed. However, using `-E` doesn't seem to work either:

```
$ echo "SOURCE bootstrap_centos_5.dockerfile" | sed -E 's|^SOURCE\s+(\S*)\s*|\1|'
SOURCE bootstrap_centos_5.dockerfile
```

Pretty sure the script wants just `bootstrap_centos_5.dockerfile` there.

Comment (technoweenie) [inline docker/docker+.bsh]: The mac's version of sed doesn't seem to like `\s` or `\w`.

```
$ echo "SOURCE bootstrap_centos_5.dockerfile" | sed -E 's|^SOURCE[ ]+([0-9a-zA-Z\.\_\-]*)|\1|'  
bootstrap_centos_5.dockerfile
```

**Comment by technoweenie** [comment]
This is working great for me now.  I definitely want to confirm that gpg signing works before I merge it though.

File: docker/README.md [hunk 3]
```diff
@@ -164,91 +233,106 @@ In order to sign packages, you need to generate and place GPG keys in the right
...
-    9. Enter a very secure password, make sure you will not forget it
+    9. Enter a secure password, make sure you will not forget it
     10. Generate Entropy!
     
-2. gpg -a --export > ./docker/public.key
-
-3. gpg -a --export-secret-keys > ./docker/signing.key
+    export --export-secret-key '<key ID>!' >
+    
+e.g. `export --export-secret-key '547CF247!' > ./docker/git-lfs_centos_7.key`
```
Comment (technoweenie) [inline docker/README.md]: Of course, this doesn't work on the mac :)

    $ gpg -a --export-secret-keys <key ID>

File: docker/gpg-agent_start.bsh [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+CUR_DIR=$(dirname ${BASH_SOURCE[0]})
+IMAGE_NAME=gpg-agent_debian_8
+
+: ${SUDO=`if ( [ ! -w /var/run/docker.sock ] && id -nG | grep -qwv docker && [ "${DOCKER_HOST:+dh}" != "dh" ] ) && which sudo > /dev/null 2>&1; then echo sudo; fi`}
+
+if [ "$(docker inspect -f {{.State.Running}} git-lfs-gpg)" != "true" ]; then
+  #Don't rebuild WHILE RUNNING
+  if [[ ${DOCKER_AUTOBUILD-1} != 0 ]]; then
+    ${CUR_DIR}/build_dockers.bsh ${IMAGE_NAME}.dockerfile
```
Comment (andyneff) [inline docker/gpg-agent_start.bsh]: This should have ${CUR_DIR} in the dockerfile path

File: docker/README.md [hunk 5]
```diff
@@ -91,72 +73,159 @@ once and saved. (This is done in CentOS by using the already existing
...
+
+In order to distribute git-lfs **and** build dependencies, the dependencies that 
+that were built for you by `build_docker.bsh` need to be saved too. Most of these
+are downloaded by yum/apt-get and do not need to be saved, but a few are not.
+In order to save the necessary dependencies, call `./docker/run_dockers.bsh` on 
+`git-lfs-full-build_*.dockerfile` before `git-lfs_*.dockerfile` and the rpms 
+will be extracted from the bootstrapped images and saved in the `./repo` directory.
+(This _can_ be done in one command. Order matters)
+
+    ./docker/run_dockers.bsh git-lfs-full-build_*.dockerfile git-lfs_*.dockerfile
```
Comment (technoweenie) [inline docker/README.md]: This command wants absolute paths to docker files. This what I had to run:

    $ ./docker/run_dockers.bsh docker/git-lfs-full-build_*.dockerfile docker/git-lfs_*.dockerfile

File: docker/README.md [hunk 6]
```diff
@@ -91,72 +73,159 @@ once and saved. (This is done in CentOS by using the already existing
...
+it automatically creates a repository for distribution too. Each distro gets a
+repo generated in `./repos/{DISTRO_NAME}/{VERSION #}`. Just drop the repo
 directory onto a webserver and you have a fully functioning Linux repo. (See
 Testing the Repositories below for more detail)
 
 The two major packages included are:
-git-lfs-....* - the git-lfs package
-git-lfs-repo-release....* - A package to install the repo.
+`git-lfs-....*` - the git-lfs package
+`git-lfs-repo-release....*` - A package to install the repo.
```
Comment (technoweenie) [inline docker/README.md]: What's the difference here? I should only be distributing the regular `git-lfs` packages, and not the repo release ones, right?

Comment (andyneff) [inline docker/README.md]: `git-lfs-....*` is the actual git-lfs package containing the git-lfs executable
`git-lfs-repo-release....*` is a package that will install a repo described by `REPO_HOSTNAME` that when hosted, can provide the user with git-lfs through their package manager program (yum, apt-get, etc...). If the user downloads `git-lfs-repo-release....*` they do not need to download `git-lfs-....*`

**Comment by technoweenie** [comment]
Thanks for this! I'm excited to try this out for v0.6.0 :)

**Pull request merged**